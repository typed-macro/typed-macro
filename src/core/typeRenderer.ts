import { mkdir, writeFile } from 'fs/promises'
import { dirname } from 'path'
import { NamespacedTypes } from './exports'

export type TypeRendererOptions = {
  types: NamespacedTypes
  typesPath: string
}

export type AppendedTypeUpdater = (type: string) => void

export type TypeRenderer = {
  render: () => string
  write: () => Promise<void>
  append: (namespace: string, type: string) => AppendedTypeUpdater
}

export function createTypeRenderer({
  types,
  typesPath,
}: TypeRendererOptions): TypeRenderer {
  return {
    render: () => renderTypes(types),
    write: async () => {
      try {
        await mkdir(dirname(typesPath), { recursive: true })
        await writeFile(typesPath, renderTypes(types))
      } catch (e) {
        throw new Error(`Error when write type declaration file: ${e}`)
      }
    },
    append: (namespace, type) => {
      let t = types[namespace]
      if (!t) {
        t = types[namespace] = {
          moduleScope: [],
          macroScope: [],
        }
      }
      const i = t.moduleScope.push(type) - 1
      return (s) => (t.moduleScope[i] = s)
    },
  }
}

export function renderTypes(types: NamespacedTypes) {
  const namespaces = Object.keys(types)
  return namespaces
    .map((ns) => {
      const item = types[ns]
      return `declare module '${ns}' {
${[item.moduleScope.join('\n'), item.macroScope.join('\n')]
  .filter((t) => !!t)
  .join('\n')}
}`
    })
    .join('\n')
}
