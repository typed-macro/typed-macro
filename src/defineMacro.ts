import { validateFnName } from '@/common'
import { Macro, MacroHandler, MacroWithType } from '@/runtime/types'

type MacroBuilder = {
  /**
   * Add custom type definition, it will be written to .d.ts before macro signature.
   */
  withCustomType: (typeDefinition: string) => Omit<MacroBuilder, 'withHandler'>
  /**
   * Add signature of macro. one macro requires at least one signature.
   * @param signature function signature, like '(s: string): void'
   * @param comment comment for the signature, will be written to .d.ts
   * before the corresponding function signature.
   */
  withSignature: (
    signature: string,
    comment?: string
  ) => Omit<MacroBuilder, 'withCustomType'>
  /**
   * Set the transform handler and get the macro.
   */
  withHandler: (handler: MacroHandler) => Macro
}

type MacroMeta = {
  signatures: {
    comment?: string
    signature: string
  }[]
  customTypes: string[]
}

/**
 * Define a macro.
 * @param name the name of the macro, should be a valid identifier name.
 * @return macro builder.
 */
export function defineMacro(name: string): Omit<MacroBuilder, 'withHandler'> {
  if (!validateFnName(name))
    throw new Error(`'${name}' is not a valid macro name!`)

  const meta: MacroMeta = {
    signatures: [],
    customTypes: [],
  }

  const builder: MacroBuilder = {
    withCustomType(typeDefinition) {
      meta.customTypes.push(typeDefinition)
      return builder
    },
    withSignature(signature, comment) {
      meta.signatures.push({ signature, comment })
      return builder
    },
    withHandler(handler) {
      if (meta.signatures.length === 0)
        throw new Error(
          `Please call .withSignature() before .withHandler() to specify at least one signature for macro '${name}'`
        )
      return {
        name,
        apply: handler,
        __types: renderMacroType(name, meta),
      } as MacroWithType
    },
  }

  return builder
}

function renderMacroType(name: string, meta: MacroMeta) {
  return [
    meta.customTypes.join('\n'),
    meta.signatures
      .map((s) =>
        s.comment
          ? `  /** ${s.comment} **/
  export function ${name}${s.signature}`
          : `  export function ${name}${s.signature}`
      )
      .join('\n'),
  ]
    .filter((t) => !!t)
    .join('\n')
}
