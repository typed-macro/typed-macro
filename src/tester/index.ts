export * from './transformer'
export * from './typeRenderer'
