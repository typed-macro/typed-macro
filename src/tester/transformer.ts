import { NamespacedMacros } from '@/core/exports'
import { createTransformer, TransformerOptions } from '@/core/transformer'

export type TestTransformerContext = {
  /**
   * Source code.
   */
  code: string

  /**
   * Path of the source file.
   *
   * Default to 'test.tsx'
   */
  filepath?: string

  /**
   * Is in SSR mode.
   *
   * Default to false.
   */
  ssr?: boolean

  /**
   * Is in dev mode.
   *
   * Default to false.
   *
   * If in dev mode, the import statements that imports macros
   * will be retained and rewritten by the runtime
   * to track the module's dependence on the macro.
   */
  dev?: boolean
}

export type TestTransformer = {
  (ctx: TestTransformerContext, macros: NamespacedMacros): string
  (code: string, macros: NamespacedMacros): string
}

/**
 * Create a transformer to help test macros.
 *
 * e.g.
 * ```typescript
 * const aMacro = defineMacro(...)
 *   .withSignature(...)
 *   .withHandler(...)
 *
 * const transform = createTestTransformer()
 *
 * expect(
 *   transform(`some code...`, { '@ns': [aMacro] })
 * ).toBe(`some code...`)
 * ```
 *
 * @param options the same as the one of transformer
 */
export function createTestTransformer(
  options: TransformerOptions = {}
): TestTransformer {
  const transform = createTransformer(options)
  return (input, macros) => {
    const {
      code,
      filepath = 'test.tsx',
      ssr = false,
      dev = false,
    } = typeof input === 'string' ? { code: input } : input
    return transform({ code, filepath, ssr, dev }, macros) || code
  }
}
