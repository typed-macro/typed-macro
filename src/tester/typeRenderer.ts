import { NamespacedExportable, normalizeExports } from '@/core/exports'
import { renderTypes } from '@/core/typeRenderer'

export type TestTypeRenderer = (exports: NamespacedExportable) => string

/**
 * Create a type renderer to help test the type of macros/modules.
 *
 * e.g.
 * ```typescript
 * const aMacro = defineMacro(...)
 *   .withSignature(...)
 *   .withHandler(...)
 *
 * const renderType = createTestTypeRenderer()
 *
 * expect(
 *   renderType({
 *     '@ns': {
 *       macros: [aMacro]
 *       customTypes: `some types...`
 *     },
 *     '@another': {
 *       code: `some code...`
 *       customTypes: `some types...`
 *     }
 *   })
 * ).toMatchSnapshot()
 * ```
 */
export function createTestTypeRenderer(): TestTypeRenderer {
  return (exports) => {
    const normalized = normalizeExports(exports)
    return renderTypes(normalized.types)
  }
}
