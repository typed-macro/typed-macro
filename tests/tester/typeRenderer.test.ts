import { createTestTypeRenderer } from '@/tester/typeRenderer'
import { defineMacro } from '@/defineMacro'

describe('createTypeRenderer()', () => {
  it('should work', () => {
    const render = createTestTypeRenderer()

    const echoMacro = defineMacro('echo')
      .withSignature('(): void')
      .withHandler(({ path }) => {
        path.remove()
      })

    expect(
      render({
        '@macro': {
          macros: [echoMacro],
        },
        '@helper': {
          code: `export const a = 1`,
          customTypes: `export const a: number`,
        },
      })
    ).toMatchSnapshot()
  })
})
