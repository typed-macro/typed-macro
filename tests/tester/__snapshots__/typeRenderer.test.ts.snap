// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTypeRenderer() should work 1`] = `
"declare module '@macro' {
  export function echo(): void
}
declare module '@helper' {
export const a: number
}"
`;
